contracts/compounder/GenericVault.sol

```diff
@@ -0,0 +1,350 @@
+            _withdrawable = _withdrawable - _penalty;
+            IStrategy(strategy).withdraw(_withdrawable);
+        }
+        return _withdrawable;
+    }
+
+    /// @notice Get the withdraw penalty amount
+    /// @param _amount Amount of asset
+    /// @return penalty amount
+    function _getWithdrawalPenalty(uint256 _amount) internal view returns (uint256) {
+        return (_amount * withdrawalPenalty) / FEE_DENOMINATOR;
+    }
+
+    /// @notice Unstake underlying token in proportion to the amount of shares sent
+    /// @param _shares - the number of shares sent
+    /// @return withdrawn - the amount of underlying returned to the user
+    function redeem(
+        uint256 _shares,
+        address _receiver,
+        address _owner
+    ) public notToZeroAddress(_receiver) notToZeroAddress(_owner) nonReentrant returns (uint256 withdrawn) {
+        // Check allowance if owner if not sender
+        if (msg.sender != _owner) {
+            uint256 currentAllowance = allowance(_owner, msg.sender);
+            require(currentAllowance >= _shares, "ERC4626: redeem exceeds allowance");
+            _approve(_owner, msg.sender, currentAllowance - _shares);
+        }
+
+        // Withdraw from extra rewards
+        for (uint256 i = 0; i < extraRewards.length; i++) {
+            IBasicRewards(extraRewards[i]).withdraw(_owner, _shares);
+        }
+
+        // Withdraw requested amount of underlying
+        uint256 _withdrawable = _withdraw(_owner, _shares);
+        // And sends back underlying to user
+        IERC20(underlying).safeTransfer(_receiver, _withdrawable);
+        emit Withdraw(msg.sender, _receiver, _owner, _withdrawable, _shares);
+        return _withdrawable;
+    }
+
+    /// @notice Claim rewards and swaps them to FXS for restaking
+    /// @dev Can be called by anyone against an incentive in FXS
+    /// @dev Harvest logic in the strategy contract
+    function harvest() public virtual {
+        uint256 _harvested = IStrategy(strategy).harvest();
+        emit Harvest(msg.sender, _harvested);
+    }
+
+    modifier notToZeroAddress(address _to) {
+        require(_to != address(0), "Invalid address!");
+        _;
+    }
+
+    /* --------------------------------------------------------------
+     * ERC20 hooks 
+    ----------------------------------------------------------------- */
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 amount
+    ) internal override {
+        // Withdraw extra rewards for the "from" address to update their earned
+        // amount when updateReward is called
+        for (uint256 i = 0; i < extraRewards.length; i++) {
+            IBasicRewards(extraRewards[i]).withdraw(from, amount);
+        }
+
+        // Stake extra rewards for the "to" address
+        for (uint256 i = 0; i < extraRewards.length; i++) {
+            IBasicRewards(extraRewards[i]).stake(to, amount);
+        }
+    }
```

Comment:
Should we make the following entry points non-reentrant for the sake of completeness ;
```transfer``` and ```transferFrom```

Reply:
I don't think it is needed on those functions.